keps/sig-scheduling/3243-respect-pod-topology-spread-after-rolling-upgrades/README.md

```diff
@@ -0,0 +1,860 @@
+`MatchLabelKeys` adds another constraint to how this group of pods is identified: 
+the scheduler will use those keys to look up label values from the incoming pod; 
+and those key-value labels are ANDed with `LabelSelector` to select the group of 
+existing pods over which spreading will be calculated.
+
+A new field named `MatchLabelKeys` will be introduced to`TopologySpreadConstraint`:
+```go
+type TopologySpreadConstraint struct {
+	MaxSkew           int32
+	TopologyKey       string
+	WhenUnsatisfiable UnsatisfiableConstraintAction
+	LabelSelector     *metav1.LabelSelector
+
+	// MatchLabelKeys is a set of pod label keys to select the pods over which 
+	// spreading will be calculated. The keys are used to lookup values from the
+	// incoming pod labels, those key-value labels are ANDed with `LabelSelector`
+	// to select the group of existing pods over which spreading will be calculated
+	// for the incoming pod. Keys that don't exist in the incoming pod labels will
+	// be ignored.
+	MatchLabelKeys []string
+}
+```
+
+Examples of use are as follows:
+```yaml
+     topologySpreadConstraints:
+        - maxSkew: 1
+          topologyKey: kubernetes.io/hostname
+          whenUnsatisfiable: DoNotSchedule
+          matchLabelKeys:
+            - app
+            - pod-template-hash
+```
+
+The scheduler plugin `PodTopologySpread` will obtain the labels from the pod 
+labels by the keys in `matchLabelKeys`. The obtained labels will be merged 
+to `labelSelector` of `topologySpreadConstraints` to filter and group pods. 
+The pods belonging to the same group will be part of the spreading in 
+`PodTopologySpread`.
+
+Finally, the feature will be guarded by a new feature flag. If the feature is 
+disabled, the field `matchLabelKeys` is preserved if it was already set in the 
+persisted Pod object, otherwise it is silently dropped; moreover, kube-scheduler 
+will ignore the field and continue to behave as before.
+
+### Test Plan
+
+<!--
+**Note:** *Not required until targeted at a release.*
+
+Consider the following in developing a test plan for this enhancement:
+- Will there be e2e and integration tests, in addition to unit tests?
+- How will it be tested in isolation vs with other components?
+
+No need to outline all of the test cases, just the general strategy. Anything
+that would count as tricky in the implementation, and anything particularly
+challenging to test, should be called out.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
+when drafting this test plan.
+
+[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
+-->
+
+- Unit and integration tests:
+  - Feature gate enable/disable tests
+  - `MatchLabelKeys` in `TopologySpreadConstraint` works as expected
+- Benchmark tests:
+  - Verify no significant performance degradation
```

Comment:
In existing benchmarks? Or new ones?

I'm assuming existing don't exercise the newly added feature - are we ok with it?

Reply:
I plan to benchmark this using the exist cases in `k8s.io/kubernetes/test/integration/scheduler_perf` to verify if there's no significant performance degradation.